rfcs/hse/0004-kvdb-kvs-introspection-apis.md

```diff
@@ -0,0 +1,156 @@
+# KVDB/KVS Introspection APIs
+
+## Problem
+
+HSE does not currently have any way to introspect KVDB/KVS information.
+
+## Requirements
+
+HSE should provide APIs for the following:
+
+- KVDB
+  - create parameters
+  - runtime parameters
+  - home
+  - KVS names (done)
+- KVS
+  - create parameters
+  - runtime parameters
+  - name
+
+Any REST APIs will support the notion of multiple KVDBs.
+
+## Non-Requirements
+
+- Will not add support for multiple KVDBs in other areas of HSE
+
+## Solution
+
+Create new APIs in C, Python, and REST in order to expose this information. REST
+APIs are necessary because you cannot introspect a KVDB open in another process.
+
+### Overview
+
+#### C APIs
+
+The following C APIs will be added behind the experimental wall:
+
+```c
+/* Should this API return an error in the case of a NULL KVDB?
+ *
+ * Returns a pointer to the kvdb_home member in ikvdb struct.
+ */
+const char *
+hse_kvdb_home_get(const struct hse_kvdb *kvdb);
+
+/* Should this API return an error in the case of a NULL KVDB?
+ *
+ * Returns a pointer to the name member in ikvs struct.
+ */
+const char *
+hse_kvs_name_get(const struct hse_kvs *kvs);
+
+/* Returns a JSON representation of the parameter value, ie the value in
+ * key=value.
+ */
+hse_err_t
+hse_kvdb_param_get(const struct hse_kvdb *kvdb, const char *param, char *buf, size_t buf_len);
+
+/* Returns a JSON representation of the parameter value, ie the value in
+ * key=value.
+ */
+hse_err_t
+hse_kvs_param_get(const struct hse_kvs *kvs, const char *param, char *buf, size_t buf_len);
+
+enum hse_mclass {
+  HSE_MCLASS_CAPACITY,
+  HSE_MCLASS_STAGING,
+};
+
+struct hse_mclass_info {
+  uint64_t total_bytes;     /**< total space in the file-system containing this media class */
+  uint64_t available_bytes; /**< available space in the file-system containing this media class */
+  uint64_t allocated_bytes; /**< allocated storage space for a media class */
+  uint64_t used_bytes;      /**< used storage space for a media class */
+  char     path[PATH_MAX];  /**< path to media class */
+};
+
+/* Populates the above storage info struct with given information.
+ *
+ * Con: Since we only track data owned by each media class, we drop the storage
+ * statistics about the kvdb.meta file, but I don't see that as being super
+ * important.
+ */
+hse_err_t
+hse_kvdb_mclass_info_get(const struct hse_kvdb *kvdb, enum hse_mclass mclass, struct hse_mclass_info *info);
+```
```

Comment:
@tristan957 mentioned the idea of using opaque structs w/ "getter" functions instead of using public struct definitions.  An API using opaque structs might look like this:
```c
hse_err_t
hse_kvdb_mclass_info_get(const struct hse_kvdb *kvdb, enum hse_mclass mclass, struct hse_mclass_info **info);

uint64_t
hse_mclass_info_get_total_bytes( struct hse_mclass_info *info );

const char *
hse_mclass_info_get_path(struct hse_mclass_info *info );

// one method to free different opaque structs
void hse_free(void *obj);
```
The advantage of this approach is that API can be easily extended while maintaining compatibility.  With the public struct, the size and layout of the struct becomes part of the API, which limits extensibility.

Reply:
If you go with hse_free(), then we should definitely remove kvs_names_free() before final tag. In fact, it might make sense to just do that now.

Reply:
@alex-tomlinson, please see https://github.com/hse-project/rfcs/pull/5#issuecomment-926034196.

I think having getter functions is unnecessary complication for a struct that's rarely used and less likely to extend in the future..

Reply:
It's a bigger question that just this struct.  It's a style choice going forward: public structs or opaque with getters.  Right now we have two structs:  mclass info and compaction status.  We will likely want to add additional structs in the future.

Reply:
If it's a style choice, then it make sense to have getters for mclass info just to be consistent with everything else.

Reply:
I was thinking that we hadn't yet leaned one way or the other (ignoring experimental APIs).  This proposed params API is getter like, but that's a special case (I think?) because: 1) it is chock full of experimental struct members that we really don't want stuck in public struct definitions, and 2) it mirrors the paramc/paramv style of name/value pairs.

Reply:
Introspection could return JSON, exchanging parsing for getters.  I'm not sure that's a good tradeoff, but it is an option.

Reply:
So I guess returning it as a protobuf is off the table then?  :-)  There are no great options.  Getters are probably the mostly commonly used approach.

Reply:
I just wanted a bit more discussion on the topic before committing to a style.  It's a C library, so using C structs is the path of least resistance.   Public structs it is.